### DISCUSSION TITLE
question about stencils

### DISCUSSION THREAD

**Discussion by alanjhewitt** [MetOffice/simulation-systems#218] [LFRic]
there is something called "tile_stencil" in jule_exp_kernel_mod.F90 that is used in the calculation of frac_surft. it is not present in the ukca equivalent kernel when calculating frac_surft.

i could do with some help getting tile_stencil into the ukca kernel. for some reason, the jules kernel does not have an associated algorithm layer

From Jules jules_exp_kernel_mod.F90

_do l = 1, land_field
      coast%fland(l) = flandg(ainfo%land_index(l),1)
      do n = 1, n_land_tile
        ! Jules requires fractions with respect to the land area
        ainfo%frac_surft(l, n) = real(tile_fraction(tile_stencil(1,1,ainfo%land_index(l))+n-1), r_um) &
             / coast%fland(l)
      end do
    end do_

From UKCA aerosol_ukca_kernel_mod.F90

    _do i = 1, n_land_pts
      do m = 1, n_land_tile
 
        frac_surft( i, m ) =                                                   &
         real( tile_fraction( map_tile( 1, land_index(i) ) + m - 1 ), r_um ) / &
         frac_land( land_index(i),1 )
 
      end do
    end do_

There is also some stencil code in the following modules (the first of which calls the jules kernel):
science/um_physics_interface/source/psy/psykal_lite_phys_mod.F90
science/gungho/source/psy/psykal_lite_mod.F90

**Comment by stevemullerworth** [comment]
It's not clear whether you need a stencil here. The Jules calculation seems to use only element 1 of the stencil (that's the second index of the tile_stencil array). Element 1 of the stencil is the central point of the stencil, so it is the same as `map_tile`. You can think of `map_tile` being a stencil of extent 1.

While stencils are not yet supported for domain kernels (hence the need for psykal-lite code), normal maps are supported.

**Reply by alanjhewitt** [reply]
I have now solved the problem. This was a red herring, the stencils were required for Jules because of halos and not for UKCA.
